src/main/scala/datalog/execution/ir/IRTreeGenerator.scala

```diff
@@ -110,6 +141,8 @@ class IRTreeGenerator(using val ctx: InterpreterContext)(using JITOptions) {
     ast match {
       case ProgramNode(ruleMap) =>
         val scc = ctx.precedenceGraph.scc(ctx.toSolve)
+        if ctx.precedenceGraph.hasNegativeCycle(ctx.storageManager.allRulesAllIndexes) then
```

Comment:
I think this check should go in the AST transform, not the IR generator, since it's a datalog semantic check